packages/@lwc/synthetic-shadow/src/3rdparty/inner-text.ts

```diff
@@ -0,0 +1,245 @@
+
+    if (selection === null) {
+        return null;
+    }
+
+    const ranges: Range[] = [];
+    for (let i = 0; i < selection.rangeCount; i++) {
+        ranges.push(selection.getRangeAt(i));
+    }
+
+    const state: SelectionState = {
+        element,
+        onselect: win.onselect,
+        onselectstart: win.onselectstart,
+        onselectionchange: win.onselectionchange,
+        ranges,
+    };
+
+    win.onselect = null;
+    win.onselectstart = null;
+    win.onselectionchange = null;
+
+    return state;
+}
+
+function restoreSelectionState(state: SelectionState | null): void {
+    if (state === null) {
+        return;
+    }
+
+    const { element, onselect, onselectstart, onselectionchange, ranges } = state;
+
+    const win = getOwnerWindow(element);
+    const selection = getWindowSelection(element)!;
+
+    selection.removeAllRanges();
+    for (let i = 0; i < ranges.length; i++) {
+        selection.addRange(ranges[i]);
+    }
+
+    win.onselect = onselect;
+    win.onselectstart = onselectstart;
+    win.onselectionchange = onselectionchange;
+}
+
+/**
+ * Gets the "innerText" of a text node using the Selection API
+ *
+ * NOTE: For performance reasons, since this function will be called multiple times while calculating the innerText of
+ *       an element, it does not restore the current selection.
+ */
+function getTextNodeInnerText(textNode: Text): string {
+    const selection = getWindowSelection(textNode);
+
+    if (selection === null) {
+        return textNode.textContent || '';
+    }
+
+    const range = document.createRange();
+    range.selectNodeContents(textNode);
+    const domRect = range.getBoundingClientRect();
+
+    if (domRect.height <= 0 || domRect.width <= 0) {
+        // the text node is not rendered
+        return '';
+    }
+
+    // Needed to remove non rendered characters from the text node.
+    selection.removeAllRanges();
+    selection.addRange(range);
+
+    return selection.toString();
+}
+
+const nodeIsElement = (node: Node): node is Element => node.nodeType === ELEMENT_NODE;
+const nodeIsText = (node: Node): node is Text => node.nodeType === TEXT_NODE;
+
+/**
+ * Spec: https://html.spec.whatwg.org/multipage/dom.html#inner-text-collection-steps
+ * One spec implementation: https://github.com/servo/servo/blob/721271dcd3c20db5ca8cf146e2b5907647afb4d6/components/layout/query.rs#L1132
+ */
+function innerTextCollectionSteps(node: Node): InnerTextItem[] {
+    const items: InnerTextItem[] = [];
+
+    if (nodeIsElement(node)) {
+        const { tagName } = node;
+        const computedStyle = getElementComputedStyle(node);
+
+        if (tagName === 'OPTION') {
+            // For options, is hard to get the "rendered" text, let's use the original getter.
+            return [1, innerTextGetter!.call(node), 1];
+        } else if (tagName === 'TEXTAREA') {
+            return [];
+        } else {
+            const childNodes = childNodesGetterPatched.call(node);
```

Comment:
why are u using this instead of just `node.childNodes`?

Reply:
mainly because of [this comment](https://github.com/salesforce/lwc/pull/2015#discussion_r480130390)

Reply:
When using the standard `node.childNodes` it is less obvious if you are using the original `Node.prototype.childNodes` or the patched version. I suggested this change because Invoking the patched version directly eliminates this issue.

Reply:
well, if that's the case, we might need to do more, consider that when going with the dot notation, it checks if node as `hasMountedChildren`, which translates to is slot or host element, but when calling the patched directly, it checks if it is a host, that seems to be related to my comment in that patched method:

```
    // nothing to do here since this does not have a synthetic shadow attached to it
    // TODO [#1636]: what about slot elements?
```

Reply:
in general, something is off!